20250404-RS-workflow-rerun-from-activity.md

```diff
@@ -0,0 +1,142 @@
+# Workflow: Rerun from Activity
+
+* Author(s): @joshvanl, @whitwaldo
+
+## Overview
+
+This proposal details the ability to rerun a workflow from a previous point in its history.
+A workflow in a terminal state can be rerun from a failed activity, before the failed activity, or at any activity in the history of a successful or failed workflow.
+
+## Background
+
+It is often the case that it's desirable to re-run business logic implemented inside a workflow.
+This could be because an activity in the workflow failed due to a transient error, an external dependency changed or a resource is now available to an activity, or it's just desirable for some subset of the last set of Activities to be rerun.
+Dapr should provide the functionality to rerun a workflow from any point in its history.
+
+## Related Items
+
+https://github.com/dapr/proposals/pull/79
+
+## Design
+
+The following proto RPC and messages will be added to durabletask, exposed via each SDK.
+This API implements rerunning a workflow from a specific Activity in the history of the workflow.
+The Activity to be rerun is chosen via its associated event ID.
+
+All `Activties` are assigned an event ID in the durabletask history.
+While `Timers` and `RaiseEvents` tasks are also assigned an event ID in the durabletask history, the workflow cannot be rerun from these events using this API.
+Not only would supporting rerunning the workflow from these two event types require a significant code refactor, in practice, users are only interested in rerunning workflow from a specific _activity_, not a "control event".
+
+It must be the case that the workflow is in a _terminal_ state before the rerun can be executed.
```

Comment:
Is there a cancelled terminal state or only succeeded and failed? Just thinking that there maybe various use cases where you want to edit an inflight workflow and effectively cancel and restart from activity X but not have all the intermidary workflow executions classified as failed. Thinking about how this is represented in other types of workflow engines like GitHub Actions. Maybe we can achieve this without an additional terminal state??

Reply:
There are [several states](https://github.com/microsoft/durabletask-dotnet/blob/main/src/Client/Core/OrchestrationRuntimeStatus.cs) that could be considered terminal for our purposes here. I read this proposal specifically as looking to avoid having someone change behavior while an activity is running (e.g. the workflow needs to be in such a state that an activity isn't going to be imminently invoked).

To that end, I'd say any of the following states would be fine then:
- Failed
- Canceled
- Terminated
- Suspended